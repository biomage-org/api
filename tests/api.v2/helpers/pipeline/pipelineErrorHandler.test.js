const { GEM2S_PROCESS_NAME } = require('../../../../src/api.v2/constants');

const pipelineErrorHandler = require('../../../../src/api.v2/helpers/pipeline/pipelineErrorHandler');
const getPipelineStatus = require('../../../../src/api.v2/helpers/pipeline/getPipelineStatus');

const sendNotification = require('../../../../src/api.v2/helpers/pipeline/hooks/sendNotification');

jest.mock('../../../../src/api.v2/helpers/pipeline/hooks/sendNotification');
jest.mock('../../../../src/api.v2/helpers/pipeline/getPipelineStatus');

const ioEmitSpy = jest.fn();

const mockExperimentId = 'mockExperimentId';
const processName = GEM2S_PROCESS_NAME;

const mockIo = {
  sockets: {
    emit: ioEmitSpy,
  },
};


const mockParsedErrorMessage = {
  taskName: 'pipelineError',
  experimentId: mockExperimentId,
  apiUrl: 'https://mock-sandbox-id.biomage.net',
  input: {
    authJWT: 'Bearer mockBearerToken',
    experimentId: mockExperimentId,
    error: 'TIMED_OUT',
    taskName: 'pipelineError',
    sandboxId: 'mock-sandbox-id',
    activityId: 'pipeline-mock-activity-id',
    processName,
  },
};

describe('pipelineErrorHandler', () => {
  it('Handles error correctly', async () => {
    await pipelineErrorHandler(mockIo, mockParsedErrorMessage);

    const [wsChannel, wsPayload] = ioEmitSpy.mock.calls[0];

    expect(ioEmitSpy).toHaveBeenCalledTimes(1);
    expect(getPipelineStatus).toHaveBeenCalledTimes(1);
    expect(getPipelineStatus).toHaveBeenCalledWith(mockExperimentId, processName);

    expect(wsChannel).toEqual(`ExperimentUpdates-${mockExperimentId}`);
    expect(wsPayload).toMatchSnapshot();

    expect(sendNotification).toHaveBeenCalledTimes(1);
    expect(sendNotification).toHaveBeenCalledWith(mockParsedErrorMessage);
  });
});
