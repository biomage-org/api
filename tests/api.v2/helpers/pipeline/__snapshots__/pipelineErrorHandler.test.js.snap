// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pipelineErrorHandler Handles error correctly 1`] = `
Object {
  "apiUrl": "https://mock-sandbox-id.biomage.net",
  "experimentId": "mockExperimentId",
  "input": Object {
    "activityId": "pipeline-mock-activity-id",
    "authJWT": "Bearer mockBearerToken",
    "error": "TIMED_OUT",
    "experimentId": "mockExperimentId",
    "processName": "gem2s",
    "sandboxId": "mock-sandbox-id",
    "taskName": "pipelineError",
  },
  "status": Object {
    "gem2s": Object {
      "completedSteps": Array [],
      "startDate": null,
      "status": "SUCCEEDED",
      "stopDate": null,
    },
  },
  "taskName": "pipelineError",
  "type": "gem2s",
}
`;
